portfolio/src/main/java/com/google/sps/servlets/HistoryCommentsServlet.java

```diff
@@ -0,0 +1,166 @@
+        JsonObject dataObj = new JsonParser().parse(test).getAsJsonObject();
+        // startIndex is already encoded after the first time. 
+        String startIndex = (dataObj.get("startIndex")).getAsString();
+        String encodedCursor = "";
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Message").addSort("timestamp", SortDirection.DESCENDING);
+        PreparedQuery preparedQuery = datastore.prepare(query);
+        FetchOptions options = FetchOptions.Builder.withLimit(4);
+        if(startIndex.equals("0")) {
+            System.out.println("Inside if");
+            Cursor cursor = preparedQuery.asQueryResultList(options).getCursor();
+            encodedCursor = cursor.toWebSafeString();
+            System.out.println("this is the first cursor: " + encodedCursor);
+        } else {
+            encodedCursor = startIndex;
+            System.out.println("this is the now old encodedCursor: " + encodedCursor);
+            System.out.println("this is the now old unnecodedCursor: " + Cursor.fromWebSafeString(encodedCursor));
+        }
+        options.startCursor(Cursor.fromWebSafeString(encodedCursor));
+       
+        // Iterable<Entity> results = preparedQuery.asIterable(options);
+        QueryResultList<Entity> resultList = preparedQuery.asQueryResultList(options);
+        String updatedEncodedCursor = resultList.getCursor().toWebSafeString();
+        System.out.println("this is the updated encodedCursor: " + updatedEncodedCursor);
+        System.out.println("resultList: " + resultList);
+        
+        messages.clear(); 
+
+        for (Entity messageEntity : resultList) {
+            long id = messageEntity.getKey().getId();
+            String messageBody = (String) messageEntity.getProperty("body"); 
+            long timestamp = (long) messageEntity.getProperty("timestamp");
+            Message newMessage = new Message(id, messageBody, timestamp);
+            messages.add(newMessage);
+        };
+        long fakeId = 000;
+        long fakeTimestamp = System.currentTimeMillis();
+        Message cursorPretendingToBeMessage = new Message(fakeId, updatedEncodedCursor, fakeTimestamp);
+        messages.add(cursorPretendingToBeMessage);
+
+        String messagesJson = convertToJsonUsingGson(messages);
+        System.out.println("messagesJson: " + messagesJson);
+
+        response.setContentType("application/json");
+        response.getWriter().println(messagesJson);
+    }
+
+
+    private String convertToJsonUsingGson(ArrayList messages) {
+        Gson gson = new Gson();
+        String json = gson.toJson(messages);
+        return json;
+    }
+
+
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        String value = request.getParameter(name);
+        if (value == null) {
+        return defaultValue;
+        }
+        return value;
+    }
```

Comment:
Same functions as those used in `DataServlet.java`. consider pulling them out into a common util class and use the util class functions instead of having the same function multiple times